### ISSUE TITLE
Add support for exception filters

### ISSUE THREAD

**Issue by tgiphil** [mosa/MOSA-Project#291]

**Comment by tgiphil** [comment]
Notes:

* I.12.4.2.5 Overview of except ion handling

* Update exception handler logic
	* needs to walk exception table for filters
	* if found, call filter
		* if filter returns true, calls preceding finally handlers and then filter handler
			* caveat: need to preserve that filter handler will need to be called
		* if false, continue walk for next filter
	* if none found, call finally handlers in order (same as before)
		
* Add IR.FilterStart & IR.FilterEnd instructions

* CIL.Endfilter transforms to move filter return register
	* Add Register FilterReturnRegister property to Mosa.Platform.x86.BaseArchitecture
	* For x86, use EDI register

**Comment by charsleysa** [comment]
Filters should be considered a variant of catches. 
In the exception table there should be a new field for the filter code offset similar to the catch code offset. 

Then when searching for catches of it is of filter type a call to the filter code is done.

The flow should be:
- try
- throw
- walk the stack to find the first protected region with a catch or filter handler
- if filter exists run filter code and store result
- if filters are run and no handlers accept the exception continue walking the stack until a handler that accepts the exception is found
- run any nested finally handlers
- run the found handler(s)

**Comment by tgiphil** [comment]
Correct. The difference between the two is the filter is called/evaluated before finally handler, and the finally handler is called before the filter exception handler.

**Comment by charsleysa** [comment]
We have to be careful to preserve the call stack while evaluating the filter, though the filter code is allowed to modify the variables of the Method that contains the filter.
